rp2040-hal/src/i2c/controller/eh1_0_async.rs

```diff
@@ -0,0 +1,306 @@
+                    } else {
+                        w.stop().disable();
+                    }
+
+                    w.cmd().read()
+                });
+                queued += 1;
+            }
+
+            if let Err(reason) = block_on!(self, check_rx_ready(), clear_rx_flags()).await {
+                abort_reason = Some(reason);
+                break;
+            }
+
+            buffer[received] = self.i2c.ic_data_cmd.read().dat().bits();
+            received += 1;
+        }
+
+        // wait for stop condition to be emitted.
+        block_on!(self, check_stop_deteced(), clear_rx_flags()).await;
+        self.i2c.ic_clr_stop_det.read().clr_stop_det();
+
+        if let Some(abort_reason) = abort_reason {
+            return Err(Error::Abort(abort_reason));
+        }
+        Ok(())
+    }
+
+    async fn non_blocking_write_internal(
+        &mut self,
+        bytes: impl IntoIterator<Item = u8>,
+        do_stop: bool,
+    ) -> Result<(), Error> {
+        let mut bytes = bytes.into_iter().peekable();
+        let mut abort_reason = None;
+        while let Some(byte) = bytes.next() {
+            if self.tx_fifo_full() {
+                // wait a bit
+                if let Err(reason) = block_on!(self, check_tx_ready(), clear_tx_flags()).await {
+                    abort_reason = Some(reason);
+                    break;
+                };
+            }
+
+            // else enqueue
+            let last = bytes.peek().is_none();
+            self.i2c.ic_data_cmd.write(|w| {
+                if do_stop && last {
+                    w.stop().enable();
+                } else {
+                    w.stop().disable();
+                }
+                unsafe { w.dat().bits(byte) }
+            });
+        }
+        if abort_reason.is_none() {
+            // wait for the tx_fifo to be emptied
+            block_on!(self, check_tx_completed(), clear_tx_flags()).await;
+            abort_reason = self.read_and_clear_abort_reason();
+        }
+
+        if abort_reason.is_some() || do_stop {
+            // If the transaction was aborted or if it completed
+            // successfully wait until the STOP condition has occured.
+            block_on!(self, check_stop_deteced(), clear_tx_flags()).await;
+            self.i2c.ic_clr_stop_det.read().clr_stop_det();
+        }
+
+        // Note the hardware issues a STOP automatically on an abort condition.
+        // Note also the hardware clears RX FIFO as well as TX on abort,
+        // because we set hwparam IC_AVOID_RX_FIFO_FLUSH_ON_TX_ABRT to 0.
+        if let Some(abort_reason) = abort_reason {
+            return Err(Error::Abort(abort_reason));
+        }
+
+        Ok(())
+    }
+
+    /// Writes to the i2c bus consuming bytes for the given iterator.
+    async fn write_iter_async<A, U>(&mut self, address: A, bytes: U) -> Result<(), Error>
+    where
+        U: IntoIterator<Item = u8>,
+        A: AddressMode + 'static + Into<u16>,
```

Comment:
I wonder if we should try to keep inherent implementations free from embedded-hal defined types. If we use them in our own API, it'll be difficult to upgrade embedded-hal later.

We did that in a few places with embedded-hal 0.2, and it'll make moving everything to embedded-hal 1.0 more difficult.

Also, we do not even support 10bit address mode yet, afaik.

Reply:
rather than the inherent method, should I impl [the crate](https://crates.io/crates/i2c-write-iter) I made to cover the `write_iter` that was removed between 0.2.7 and 1.0 ?

Reply:
Doesn't that complicate things further? It would mean that users of those methods would need to depend on your crate as well and import the traits, before they can call the functions.

Reply:
Indeed. I wish it hadn't been removed from the embedded-hal in the first place :/

I'll make this driver impl the trait and keep this dependency feature-gated.